README.md
# fast-rng: Fast random number generator for various distributions

[![Download](https://api.bintray.com/packages/komiya-atsushi/maven/fast-rng/images/download.svg) ](https://bintray.com/komiya-atsushi/maven/fast-rng/_latestVersion)

# Getting started

## Resolving artifacts using Maven

```xml
<repositories>
  <repository>
    <id>bintray-komiya-atsushi-maven</id>
    <url>http://dl.bintray.com/komiya-atsushi/maven</url>
  </repository>
</repositories>

<dependencies>
  <dependency>
    <groupId>biz.k11i</groupId>
    <artifactId>fast-rng</artifactId>
    <version>0.2.0</version>
  </dependency>
</dependencies>
```

## Resolving artifacts using Gradle

```groovy
repositories {
    maven {
        url  "http://dl.bintray.com/komiya-atsushi/maven" 
    }
}

dependencies {
    compile group: 'biz.k11i', name: 'fast-rng', version: '0.2.0'
}
```

## Generating random values

```java
package biz.k11i.rng.demo;

import java.util.Random;

public class FastRngDemo {
    public static void main(String[] args) {
        // Fast-rng requires a java.util.Random instance to generate
        // uniformly distributed random values.
        Random random = new Random();

        System.out.println(
                // Generate a gaussian random value.
                GaussianRNG.FAST_RNG.generate(random)
        );
    }
}
```


# Supported distributions

- [Gaussian distribution (Normal distribution)](https://en.wikipedia.org/wiki/Normal_distribution)
- [Exponential distribution](https://en.wikipedia.org/wiki/Exponential_distribution)
- [Gamma distribution](https://en.wikipedia.org/wiki/Gamma_distribution)
- [Beta distribution](https://en.wikipedia.org/wiki/Beta_distribution)


# License

This software is licensed under the terms of the MIT license. See LICENSE.


# Acknowledgments

## commons-math3

This product includes software developed at
The Apache Software Foundation (http://www.apache.org/).
https://github.com/apache/commons-math

## Jafama

This product includes software developed by Jeff Hain.
https://github.com/jeffhain/jafama


# References

- Tesuaki Yotsuji.
  *計算機シミュレーションのための確率分布乱数生成法.*
  Pleiades Publishing Co.,Ltd. (2010)
- Marsaglia, George, and Wai Wan Tsang.
  *The ziggurat method for generating random variables.*
  Journal of statistical software 5.8 (2000): 1-7.
- Ahrens, Joachim H., and Ulrich Dieter.
  *Computer methods for sampling from gamma, beta, poisson and bionomial distributions.*
  Computing 12.3 (1974): 223-246.
- Marsaglia, George, and Wai Wan Tsang.
  *A simple method for generating gamma variables.*
  ACM Transactions on Mathematical Software (TOMS) 26.3 (2000): 363-372.
- Best, D. J.
  *A note on gamma variate generators with shape parameter less than unity.*
  Computing 30.2 (1983): 185-188.
- Wilson, Edwin B., and Margaret M. Hilferty.
  *The distribution of chi-square.*
  Proceedings of the National Academy of Sciences 17.12 (1931): 684-688.
- Jöhnk, M. D.
  *Erzeugung von betaverteilten und gammaverteilten Zufallszahlen.*
  Metrika 8.1 (1964): 5-15.
- Sakasegawa, H.
  *Stratified rejection and squeeze method for generating beta random numbers.*
  Annals of the Institute of Statistical Mathematics 35.1 (1983): 291-302.
- Lemire, Daniel.
  *Fast random integer generation in an interval.*
  ACM Transactions on Modeling and Computer Simulation (TOMACS) 29.1 (2019): 3.


RELEASE-NOTES.md
# Release notes

## 0.2.0

- Fast-rng now requires Java 8 or later.
- Introduces a new class [`UniformRNGUtils`](fast-rng/src/main/java/biz/k11i/rng/UniformRNGUtils.java).
    - The method `UniformRNGUtils#nextInt(java.util.Random random, int bound)` is faster than [`Random#nextInt(int bound)`](https://docs.oracle.com/javase/8/docs/api/java/util/Random.html#nextInt-int-).

## 0.1.5

- Support beta random number generator

## 0.1.4

- Improve speed performance of `GammaRNG`

## 0.1.3

- Support exponential random number generator

## 0.1.2

- Provide two implementations of Ziggurat algorithm for Gaussian RNG
    - `GaussianRNG.FAST_RNG` and `GaussianRNG.GENERAL_RNG`

## 0.1.1

- Support gamma random number generator


## 0.1.0

- Initial release
- Support gaussian random number generator


fast-rng/src/main/java/biz/k11i/rng/GammaRNG.java
package biz.k11i.rng;

import java.util.Random;

import static biz.k11i.util.MathFunctions.*;
import static java.lang.Math.sqrt;

/**
 * Gamma random number generator.
 */
public interface GammaRNG {
    GammaRNG FAST_RNG = new FastRNG();
    GammaRNG GENERAL_RNG = new GeneralRNG();

    /**
     * Generates a random value sampled from gamma distribution.
     *
     * @param random random number generator
     * @param shape  shape parameter (alpha)
     * @param scale  scale parameter (beta)
     * @return a random value
     */
    double generate(Random random, double shape, double scale);

    class FastRNG implements GammaRNG {
        private final GammaRNG mt = new GammaRNGAlgorithms.MarsagliaTsang(GaussianRNG.FAST_RNG);
        private final GammaRNG exponential = new GammaRNGAlgorithms.Exponential(ExponentialRNG.FAST_RNG);
        private final GammaRNG wh = new GammaRNGAlgorithms.WilsonHilfertyApproximation(GaussianRNG.FAST_RNG);

        @Override
        public double generate(Random random, double shape, double scale) {
            if (shape >= 50) {
                return wh.generate(random, shape, scale);
            }
            if (shape != 1.0) {
                return mt.generate(random, shape, scale);
            }

            // shape == 1.0
            return exponential.generate(random, shape, scale);
        }
    }

    class GeneralRNG implements GammaRNG {
        private final GammaRNG best = new GammaRNGAlgorithms.Best();
        private final GammaRNG mt = new GammaRNGAlgorithms.MarsagliaTsang(GaussianRNG.GENERAL_RNG);
        private final GammaRNG exponential = new GammaRNGAlgorithms.Exponential(ExponentialRNG.GENERAL_RNG);
        private final GammaRNG wh = new GammaRNGAlgorithms.WilsonHilfertyApproximation(GaussianRNG.GENERAL_RNG);

        @Override
        public double generate(Random random, double shape, double scale) {
            if (shape >= 50) {
                return wh.generate(random, shape, scale);
            }
            if (shape < 0.1) {
                return best.generate(random, shape, scale);
            }
            if (shape != 1.0) {
                return mt.generate(random, shape, scale);
            }

            // shape == 1.0
            return exponential.generate(random, shape, scale);
        }
    }
}

class GammaRNGAlgorithms {
    abstract static class BaseGammaRNG implements GammaRNG {
        @Override
        public double generate(Random random, double shape, double scale) {
            return generate(random, shape) * scale;
        }

        abstract double generate(Random random, double shape);

        @Override
        public String toString() {
            return String.format("%s", this.getClass().getSimpleName());
        }
    }

    /**
     * Implementation of Gamma random number generator using Best's algorithm (1983).
     * <p>
     * Best, D. J.
     * <i>“A note on gamma variate generators with shape parameter less than unity.”</i>
     * Computing 30.2 (1983): 185-188.
     * </p>
     */
    static class Best extends BaseGammaRNG {
        @Override
        double generate(Random random, double shape) {
            double c1 = 0.07 + 0.75 * sqrt(1 - shape);
            double c2 = 1 + shape * exp(-c1) / c1;
            double c3 = 1.0 / shape;

            while (true) {
                double u1 = random.nextDouble();
                double u2 = random.nextDouble();
                double v = c2 * u1;

                if (v <= 1) {
                    double x = c1 * pow(v, c3);
                    if (u2 <= (2 - x) / (2 + x) || u2 <= exp(-x)) {
                        return x;
                    }
                } else {
                    double x = -log(c1 * c3 * (c2 - v));
                    double y = x / c1;
                    if (u2 * (shape + y - shape * y) <= 1 || u2 < pow(y, shape - 1)) {
                        return x;
                    }
                }
            }
        }
    }

    static class Exponential implements GammaRNG {
        private final ExponentialRNG exponentialRNG;

        Exponential(ExponentialRNG exponentialRNG) {
            this.exponentialRNG = exponentialRNG;
        }

        @Override
        public double generate(Random random, double shape, double scale) {
            return exponentialRNG.generate(random, 1.0) * scale;
        }

        @Override
        public String toString() {
            return String.format("Exponential[%s]", exponentialRNG.getClass().getSimpleName());
        }
    }

    /**
     * Implementation of Gamma random number generator using Marsaglia and Tsang's algorithm (2000).
     * <p>
     * Marsaglia, George, and Wai Wan Tsang.
     * <i>"A simple method for generating gamma variables."</i>
     * ACM Transactions on Mathematical Software (TOMS) 26.3 (2000): 363-372.
     * </p>
     */
    static class MarsagliaTsang extends BaseGammaRNG {
        private final GaussianRNG gaussianRNG;

        MarsagliaTsang(GaussianRNG gaussianRNG) {
            this.gaussianRNG = gaussianRNG;
        }

        @Override
        double generate(Random random, double shape) {
            if (shape >= 1) {
                return generateMT(random, shape);
            }

            double r = generateMT(random, shape + 1);
            double u = random.nextDouble();

            if (shape != 0.5) {
                return r * pow(u, 1.0 / shape);
            }

            // shape == 0.5
            return r * u * u;
        }

        double generateMT(Random random, double shape) {
            double d = shape - 1.0 / 3;
            double c = 1 / sqrt(9 * d);

            while (true) {
                double x = gaussianRNG.generate(random);
                double v = 1 + c * x;
                if (v <= 0) {
                    continue;
                }

                v = v * v * v;
                x = x * x;

                double u = random.nextDouble();
                if (u < 1 - 0.0331 * x * x) {
                    return d * v;
                }

                if (log(u) < 0.5 * x + d * (1 - v + log(v))) {
                    return d * v;
                }
            }
        }

        @Override
        public String toString() {
            return String.format("%s[%s]",
                    getClass().getSimpleName(),
                    this.gaussianRNG.getClass().getSimpleName());
        }
    }

    /**
     * Implementation of Gamma random number generator using Wilson-Hilferty approximation
     * for large shape parameter (>= 50).
     * <p>
     * Wilson, Edwin B., and Margaret M. Hilferty.
     * <i>“The distribution of chi-square.”</i>
     * Proceedings of the National Academy of Sciences 17.12 (1931): 684-688.
     * </p>
     */
    static class WilsonHilfertyApproximation extends BaseGammaRNG {
        private final GaussianRNG gaussianRNG;

        WilsonHilfertyApproximation(GaussianRNG gaussianRNG) {
            this.gaussianRNG = gaussianRNG;
        }

        @Override
        double generate(Random random, double shape) {
            double t0 = 1.0 / (9.0 * shape);
            double t1 = 1.0 - t0;
            double t2 = sqrt(t0);

            while (true) {
                double t = t1 + t2 * gaussianRNG.generate(random);
                if (t <= 0) {
                    continue;
                }

                return shape * t * t * t;
            }
        }

        @Override
        public String toString() {
            return String.format("WilsonHilfertyApproximation[%s]", gaussianRNG.getClass().getSimpleName());
        }
    }
}